### PR TITLE
Dev hyperband

### PR DESCRIPTION
**Pull request by QuanluZhang**
support advisor which is the integration of tuner and assessor;
support hyperband algorithm acting as an advisor.

### CODE REVIEW

File: docs/howto_2_CustomizedTuner.md [hunk 1]
```diff
@@ -93,3 +93,6 @@ More detail example you could see:
 > * [evolution-tuner](../src/sdk/pynni/nni/evolution_tuner)
 > * [hyperopt-tuner](../src/sdk/pynni/nni/hyperopt_tuner)
 > * [evolution-based-customized-tuner](../examples/tuners/ga_customer_tuner)
+
+## Write a more advanced automl algorithm
+The methods above are usually enough to write a general tuner. However, users may also want more methods, for example, the methods in assessor (e.g., `assess_trial`, `trial_end`), in order to have a more powerful automl algorithm. Therefore, we have another concept called `advisor` which directly inherits from `MsgDispatcherBase` in [`src/sdk/pynni/nni/msg_dispatcher_base.py`](../src/sdk/pynni/nni/msg_dispatcher_base.py). Please refer to [here](howto_3_CustomizedAdvisor) for how to write a customized advisor.
```
Comment (xuehui1991) [inline docs/howto_2_CustomizedTuner.md]: could you explain more detail here? when we need to use 'advisor'??

File: docs/howto_2_CustomizedTuner.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ So, if user want to implement a customized Tuner, she/he only need to:
 
 1) Inherit a tuner of a base Tuner class
 2) Implement receive_trial_result and generate_parameter function
-3) Write a script to run Tuner
+3) Configure your customized tuner in experiment yaml config file
```
Comment (xuehui1991) [inline docs/howto_2_CustomizedTuner.md]: why this file name("howto_2_CustomizedTuner") is so strange....

File: docs/howto_3_CustomizedAdvisor.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# **How To** - Customize Your Own Advisor
+
+*Advisor targets the scenario that the automl algorithm wants the methods of both tuner and assessor. Advisor is similar to tuner on that it receives trial configuration request, final results, and generate trial configurations. Also, it is similar to assessor on that it receives intermediate results, trial's end state, and could send trial kill command. Note that, if you use Advisor, tuner and assessor are not allowed to be used at the same time.*
+
+So, if user want to implement a customized Advisor, she/he only need to:
+
+1) Define an Advisor inheriting from the MsgDispatcherBase class
+2) Implement the handle_xxx methods
```
Comment (xuehui1991) [inline docs/howto_3_CustomizedAdvisor.md]: handle_xxx methods from MsgDispathcherBase?

File: examples/trials/mnist-hyperband/mnist.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        # Reshape to use within a convolutional neural net.
+        # Last dimension is for "features" - there is only one here, since images are
+        # grayscale -- it would be 3 for an RGB image, 4 for RGBA, etc.
+        with tf.name_scope('reshape'):
+            try:
+                input_dim = int(math.sqrt(self.x_dim))
+            except:
+                print(
+                    'input dim cannot be sqrt and reshape. input dim: ' + str(self.x_dim))
+                logger.debug(
```
Comment (yds05) [inline examples/trials/mnist-hyperband/mnist.py]: why duplicate the error message both into stdout and log?

File: docs/howto_3_CustomizedAdvisor.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+```yaml
+advisor:
+  codeDir: /home/abc/myadvisor
+  classFileName: my_customized_advisor.py
+  className: CustomizedAdvisor
+  # Any parameter need to pass to your advisor class __init__ constructor
+  # can be specified in this optional classArgs field, for example 
+  classArgs:
+    arg1: value1
```
Comment (xuehui1991) [inline docs/howto_3_CustomizedAdvisor.md]: could I dynamic change the parameters in here and I don't have to change the code in nni?  for example, advisor1 have parameter arg1 and  advisor1 have parameter arg2?

File: examples/trials/mnist-hyperband/config.yml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+#choice: local, remote, pai
+trainingServicePlatform: local
+searchSpacePath: search_space.json
+#choice: true, false
+useAnnotation: false
+advisor:
+  #choice: Hyperband
+  builtinAdvisorName: Hyperband
+  classArgs:
+    #R: the maximum STEPS
```
Comment (xuehui1991) [inline examples/trials/mnist-hyperband/config.yml]: maximum step for what?

File: examples/trials/mnist-hyperband/mnist.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+if __name__ == '__main__':
+    try:
+        # get parameters form tuner
+        RCV_PARAMS = nni.get_next_parameter()
+        logger.debug(RCV_PARAMS)
+        # run
+        params = generate_default_params()
+        params.update(RCV_PARAMS)
+        params['batch_num'] = RCV_PARAMS['STEPS'] * 10
```
Comment (xuehui1991) [inline examples/trials/mnist-hyperband/mnist.py]: why times 10 here?

File: examples/trials/mnist-hyperband/mnist.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
+"""A deep MNIST classifier using convolutional layers."""
```
Comment (xuehui1991) [inline examples/trials/mnist-hyperband/mnist.py]: I can't see much difference with old-mnist example. Could we reuse the old one and add a config for hyperband?

File: docs/howto_2_CustomizedTuner.md (hunk 1 above)
Comment (QuanluZhang) [reply docs/howto_2_CustomizedTuner.md]: sure.

File: docs/howto_2_CustomizedTuner.md (hunk 2 above)
Comment (QuanluZhang) [reply docs/howto_2_CustomizedTuner.md]: renamed by @scarlett2018

File: docs/howto_3_CustomizedAdvisor.md (hunk 3 above)
Comment (QuanluZhang) [reply docs/howto_3_CustomizedAdvisor.md]: ok, I can list all the methods.

File: docs/howto_3_CustomizedAdvisor.md (hunk 5 above)
Comment (QuanluZhang) [reply docs/howto_3_CustomizedAdvisor.md]: advisor2? and what do you mean?

File: examples/trials/mnist-hyperband/config.yml (hunk 6 above)
Comment (QuanluZhang) [reply examples/trials/mnist-hyperband/config.yml]: this is not straightforward, please refer to our doc of hyperband

File: examples/trials/mnist-hyperband/mnist.py (hunk 7 above)
Comment (QuanluZhang) [reply examples/trials/mnist-hyperband/mnist.py]: STEP controls how many resource allocated to a configuration, the minimum number of STEPS is 1 which is too small for the training.

File: examples/trials/mnist-hyperband/mnist.py (hunk 8 above)
Comment (QuanluZhang) [reply examples/trials/mnist-hyperband/mnist.py]: no. the STEPS in hyperparameters is very important.

File: examples/trials/mnist-hyperband/mnist.py (hunk 4 above)
Comment (QuanluZhang) [reply examples/trials/mnist-hyperband/mnist.py]: copied from example/trials/mnist/mnist.py, maybe better to keep it and fix this kind of problems in all the examples latter :)

File: docs/howto_2_CustomizedTuner.md (hunk 1 above)
Comment (QuanluZhang) [reply docs/howto_2_CustomizedTuner.md]: fixed, thx

File: docs/howto_3_CustomizedAdvisor.md (hunk 3 above)
Comment (QuanluZhang) [reply docs/howto_3_CustomizedAdvisor.md]: fixed, thx

File: examples/trials/mnist-hyperband/config.yml (hunk 6 above)
Comment (QuanluZhang) [reply examples/trials/mnist-hyperband/config.yml]: fixed, thx

File: examples/trials/mnist-hyperband/mnist.py (hunk 7 above)
Comment (xuehui1991) [reply examples/trials/mnist-hyperband/mnist.py]: I am little confuse. the default parameter and search space has no ‘STEP’，so what is 'STEP' come from?

**Pull request merged**